new-log-viewer/src/contexts/UrlContextProvider.tsx

```diff
@@ -0,0 +1,109 @@
+import {
+    createContext,
+    useEffect,
+    useState,
+} from "react";
+
+
+interface UrlContextType {
+    setSearchParamSet: (searchParamSet: Record<string, string | null>) => void;
+    setHashParamSet: (hashParamSet: Record<string, string | null>) => void;
+    copyToClipboard: (
+        searchParamSet: Record<string, string | null>,
+        hashParamSet: Record<string, string | null>
+    ) => void;
+}
+
+const UrlContext = createContext <UrlContextType>({} as UrlContextType);
+
+
+interface UrlContextProviderProps {
+    children: React.ReactNode
+}
+
+/**
+ * Provides a context for managing URL parameters and hash values,
+ * including utilities for setting search and hash parameters,
+ * and copying the current URL with these parameters to the clipboard.
+ *
+ * @param children.children
+ * @param children The child components that will have access to the context.
+ */
+const UrlContextProvider = ({children}: UrlContextProviderProps) => {
+    const [hashParam, setHashParam] = useState<string>(window.location.hash.substring(1));
+    useEffect(() => {
+        setHashParam(window.location.hash.substring(1));
+    }, [window.location.hash]);
```

Comment:
This may not really work unless the UI re-renders, right?
I believe a more proper way is to write something like this:
```
    useEffect(() => {
        const handleHashChange = () => {
            ...
        };

        window.addEventListener("hashchange", handleHashChange);

        return () => {
            window.removeEventListener("hashchange", handleHashChange);
        };
    }, []);

```